doc/dev-guide/src/encoding/types.md

```diff
@@ -0,0 +1,128 @@
+# Types
+
+This section describes how Rust types are encoded into Viper. 
+
+## Primitive types
+
+A small number of Rust types have direct analogues in the Viper type system. However, because it is possible to reference local variables and arguments in Rust, all variables are encoded as `Ref`s, and specific fields are used to refer to the "primitive value" contained in those `Ref`s.
+
+### `Bool`
+
+`Bool` is encoded to Viper `Bool`.
+
+```viper
+field val_bool: Bool
+predicate bool(self: Ref) {
+  acc(self.val_bool)
+}
+```
+
+### `i*`, `u*`, `char`
+
+All Rust integers (`i8`, `i16`, `i32`, `i64`, `i128`, `isize`, `u8`, `u16`, `u32`, `u64`, `u128`, `usize`) and `char` are encoded to Viper `Int`.
+
+```viper
+field val_i32: Int
+predicate i32(self: Ref) {
+  acc(self.val_i32)
+}
+```
+
+When the [`CHECK_BINARY_OPERATIONS`](../config/flags.md#check_binary_operations) flag is enabled, integer bounds are encoded with the type.
+
+```viper
+// with CHECK_BINARY_OPERATIONS
+predicate i32(self: Ref) {
+  acc(self.val_i32) && -2147483648 <= self.val_i32 && self.val_i32 <= 2147483647
+}
+```
+
+When the [`ENCODE_UNSIGNED_NUM_CONSTRAINT`](../config/flags.md#encode_unsigned_num_constraint) flag is enabled (even when `CHECK_BINARY_OPERATIONS` is not), non-negativity of unisnged integers is always encoded.
+
+```viper
+// with ENCODE_UNSIGNED_NUM_CONSTRAINT
+predicate u32(self: Ref) {
+  acc(self.val_u32) && 0 <= self.val_u32
+}
+```
+
+## Composite types
+
+There are no classes in Viper (every `Ref` instance has all fields), so non-primitive Rust types are encoded using [predicates](http://viper.ethz.ch/tutorial/?page=1&section=#predicates) and `Ref`s.
```

Comment:
This is correct, but we somewhat have (or will have) two representations for composite types:

1. a `Ref` variable and a predicate specifying all the fields, and 
2. a "snapshot", which essentially encodes the composite type in a deep value using Viper domains.

The second representation is independent of the Viper heap; it can thus, for example, be returned in a function, which is not possible using the first approach.